From martyn.welch@gefanuc.com  Thu Aug 13 16:54:34 2009
From: Martyn Welch <martyn.welch@gefanuc.com>
Date: Tue, 11 Aug 2009 13:50:49 +0100
Subject: Staging: vme: Add syncronize interrupts before removing callback
To: gregkh@suse.de
Cc: devel@linuxdriverproject.org, linux-kernel@vger.kernel.org, jirislaby@gmail.com
Message-ID: <20090811124851.1722.28998.stgit@ES-J7S4D2J.amer.consind.ge.com>


As identified by Jiri, there is no syncronisation before callback is removed.

Signed-off-by: Martyn Welch <martyn.welch@gefanuc.com>
Cc: Jiri Slaby <jirislaby@gmail.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/staging/vme/bridges/vme_tsi148.c |    5 +++++
 1 file changed, 5 insertions(+)

--- a/drivers/staging/vme/bridges/vme_tsi148.c
+++ b/drivers/staging/vme/bridges/vme_tsi148.c
@@ -480,6 +480,7 @@ int tsi148_request_irq(int level, int st
 void tsi148_free_irq(int level, int statid)
 {
 	u32 tmp;
+	struct pci_dev *pdev;
 
 	/* Get semaphore */
 	down(&(vme_irq));
@@ -495,6 +496,10 @@ void tsi148_free_irq(int level, int stat
 		tmp = ioread32be(tsi148_bridge->base + TSI148_LCSR_INTEO);
 		tmp &= ~TSI148_LCSR_INTEO_IRQEO[level - 1];
 		iowrite32be(tmp, tsi148_bridge->base + TSI148_LCSR_INTEO);
+
+		pdev = container_of(tsi148_bridge->parent, struct pci_dev, dev);
+
+		synchronize_irq(pdev->irq);
 	}
 
 	tsi148_bridge->irq[level - 1].callback[statid].func = NULL;
